rest-api/src/main/java/com/example/restapi/controllers/BookmarkController.java

```diff
@@ -0,0 +1,101 @@
+package com.example.restapi.controllers;
+
+import com.example.restapi.entities.Bookmark;
+import com.example.restapi.repositories.BookmarkRepository;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.java.Log;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+@RestController
+@RequestMapping("/api/v1/bookmarks")
+@RequiredArgsConstructor
+@Log
+public class BookmarkController {
+
+  private final BookmarkRepository bookmarkRepository;
+
+  @GetMapping
+  ResponseEntity<List<Bookmark>> fetchBookmarks() {
+    List<Bookmark> bookmarkList = new ArrayList<>();
+    bookmarkRepository.findAll().iterator().forEachRemaining(bookmarkList::add);
+    log.info("fetched bookmarks: " + bookmarkList);
+    return ResponseEntity.ok(bookmarkList);
+  }
+
+  @GetMapping("/{bookmarkId}")
+  ResponseEntity<Bookmark> fetchBookmarkById(@PathVariable String bookmarkId) {
+    log.info("fetching bookmark: " + bookmarkId);
+    return bookmarkRepository
+        .findById(bookmarkId)
+        .map(ResponseEntity::ok)
+        .orElseGet(() -> ResponseEntity.notFound().build());
+  }
+
+  @PostMapping(
+      consumes = MediaType.APPLICATION_JSON_VALUE,
+      produces = MediaType.APPLICATION_JSON_VALUE)
+  ResponseEntity<Bookmark> addBookmark(@RequestBody Bookmark bookmarkPayload) {
+    log.info("payload received: " + bookmarkPayload);
+
+    Bookmark addedBookmark = bookmarkRepository.save(bookmarkPayload);
+    log.info("added bookmark: " + addedBookmark);
+
+    return ResponseEntity.created(URI.create("/" + addedBookmark.getId())).body(addedBookmark);
+  }
+
+  @PutMapping(
+      value = "/{bookmarkId}",
+      consumes = MediaType.APPLICATION_JSON_VALUE,
+      produces = MediaType.APPLICATION_JSON_VALUE)
+  ResponseEntity<Bookmark> updateCompleteBookmark(
+      @PathVariable String bookmarkId, @RequestBody Bookmark bookmarkPayload) {
+    log.info("payload received: " + bookmarkPayload);
+
+    Optional<Bookmark> bookmarkOptional = bookmarkRepository.findById(bookmarkId);
+
+    if (bookmarkOptional.isEmpty()) return ResponseEntity.notFound().build();
+
+    Bookmark bookmark = bookmarkOptional.get();
```

Comment:
I'll accept it for now but I'll ask you to move that kind of logic to another layer like `BookmarksService`.

Reply:
Yeah, agreed @mbnsky .